### PR TITLE
olly trace: ingest custom events

### PR DESCRIPTION
**Pull request by TheLortex**
In 5.1 programmers will be able to add other runtime events. This PR modifies `olly trace` so that it includes custom events of common types (unit, int, span).

Conditional compilation was used to make it work on 5.0.

### CODE REVIEW

**Comment by kayceesrk** [comment]
How did you test this? With 5.1.0~alpha2, `opam pin .` gives me the following error building the dependent package `core`:

```bash
[ERROR] The compilation of core.v0.16.0 failed at "dune build -p core -j 7".

#=== ERROR while compiling core.v0.16.0 =======================================#
# context     2.1.3 | macos/arm64 | ocaml-base-compiler.5.1.0~alpha2 | https://opam.ocaml.org#7b30d632
# path        ~/.opam/5.1.0~alpha2/.opam-switch/build/core.v0.16.0
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p core -j 7
# exit-code   1
# env-file    ~/.opam/log/core-97555-cfb0d4.env
# output-file ~/.opam/log/core-97555-cfb0d4.out
### output ###
# should be applied to '()'; using '(struct end)' is deprecated.
# [...]
# File "_none_", line 1:
# Warning 73 [generative-application-expects-unit]: A generative functor
# should be applied to '()'; using '(struct end)' is deprecated.
# 
# File "_none_", line 1:
# Warning 73 [generative-application-expects-unit]: A generative functor
# should be applied to '()'; using '(struct end)' is deprecated.
# 
# File "_none_", line 1:
# Warning 73 [generative-application-expects-unit]: A generative functor
# should be applied to '()'; using '(struct end)' is deprecated.



<><> Error report <><><><><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+- The following actions failed
| - build core v0.16.0
+- 
- No changes have been performed
```

**Comment by Sudha247** [comment]
This failure on arm64 and mac is odd -
```
Uncaught exception:
  
  (Invalid_argument
   "Float.iround_nearest_exn: argument (nan) is too small or NaN")
```

FWIW, I was seeing the same error on #16 as well, it wasn't introduced by this PR. I have access to an arm64 machine, will try to reproduce this locally.

**Comment by TheLortex** [comment]
@kayceesrk I didn't test the 5.1 alpha so no idea why core is not building there.

My tests were done on a 5.0 compiler with the custom events patch: https://github.com/TheLortex/ocaml/tree/custom-events-5.0-rc1.

**Comment by Sudha247** [comment]
@kayceesrk I don't see any errors on `5.1.0~beta1`. Maybe we should activate a GH Actions job for 5.1.

**Comment by kayceesrk** [comment]
> Maybe we should activate a GH Actions job for 5.1.

Sounds good to me.

**Comment by kayceesrk** [comment]
Is this PR good to be merged @TheLortex @Sudha247?

File: bin/olly.ml [hunk 1]
```diff
@@ -191,9 +193,11 @@ let trace format trace_filename exec_args =
           ~name:(Runtime_events.runtime_phase_name phase)
           ~time:(ts |> ts_to_int |> int_to_span)
       in
+      let extra = Olly_custom_events.v trace_file doms in
       let init () = () in
       let cleanup () = Trace.close trace_file in
-      olly ~runtime_begin ~runtime_end ~init ~lifecycle ~cleanup exec_args
+      olly ~extra ~runtime_begin ~runtime_end ~init ~lifecycle ~cleanup
```
Comment (Sudha247) [inline bin/olly.ml:199]: Did not do a full review. Do we want to add custom events to json as well?

Comment (kayceesrk) [reply bin/olly.ml:199]: For completeness sake, we should.

**Pull request merged**